agent.go

```diff
@@ -0,0 +1,461 @@
+//
+// Copyright (c) 2017 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/yamux"
+	pb "github.com/hyperhq/runv/hyperstart/api/grpc"
+	"github.com/mdlayher/vsock"
+	"github.com/opencontainers/runc/libcontainer"
+	"github.com/opencontainers/runc/libcontainer/configs"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+type process struct {
+	id          string
+	process     libcontainer.Process
+	stdin       *os.File
+	stdout      *os.File
+	stderr      *os.File
+	consoleSock *os.File
+	termMaster  *os.File
+}
+
+type container struct {
+	sync.RWMutex
+
+	id          string
+	initProcId  string
+	container   libcontainer.Container
+	config      configs.Config
+	processes   map[string]*process
+	wgProcesses sync.WaitGroup
+	mounts      []string
+}
+
+type pod struct {
+	sync.RWMutex
+
+	id           string
+	running      bool
+	containers   map[string]*container
+	channel      channel
+	network      network
+	shareDir     string
+	wg           sync.WaitGroup
+	grpcListener net.Listener
+}
+
+const (
+	agentName       = "kata-agent"
+	exitSuccess     = 0
+	exitFailure     = 1
+	fileMode0750    = 0750
+	defaultLogLevel = logrus.InfoLevel
+)
+
+const (
+	gRPCSockPath   = "/var/run/kata-containers/grpc.sock"
+	gRPCSockScheme = "unix"
```

Comment:
I don't think these are needed. unix sock scheme proxy is only needed in the host side since proxy and shim/runtime are different processes. The origin design doesn't include it.

```
 |----runv------|----proxy-------|-------hyperstart
 container 1 ---|--------\             /---
 container 2 ---|--------->------|----<----  hyperstart actions
 exec 1---------|--------/             \---
  grpc      unixsock  yamux   serial   yamux     grpc
```
(@gnawux had drawn a picture from this ascii)

yamux.Server()'s return value implements net.Listener. So it can be directly used with grpc(grpcServer.Serve()).

However, I haven't tried. Cloud you have a try please?

Reply:
Oh yes good point, I haven't realized that. It's going to simplify the code.